src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -117,17 +123,18 @@ public GeneratorDriver TryApplyEdits(Compilation compilation, out Compilation ou
 
         public GeneratorDriver AddGenerators(ImmutableArray<ISourceGenerator> generators)
         {
-            var newState = _state.With(generators: _state.Generators.AddRange(generators));
+            // set editsFailed true, as we won't be able to apply edits with a new generator
+            var newState = _state.With(generators: _state.Generators.AddRange(generators), editsFailed: true);
             return FromState(newState);
         }
 
         public GeneratorDriver RemoveGenerators(ImmutableArray<ISourceGenerator> generators)
         {
-            var newState = _state.With(generators: _state.Generators.RemoveRange(generators));
+            var newState = _state.With(generators: _state.Generators.RemoveRange(generators), generatorStates: _state.GeneratorStates.RemoveRange(generators));
             return FromState(newState);
         }
 
-        public GeneratorDriver WithAdditionalTexts(ImmutableArray<AdditionalText> additionalTexts)
+        public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additionalTexts)
```

Comment:
Why the name change? Is this mutating now?

Reply:
Not mutating, but always returned a new generator with extra texts. Name change is just consistency

---
In reply to: [392522980](https://github.com/dotnet/roslyn/pull/42359#discussion_r392522980) [](ancestors = 392522980)

Reply:
I'd prefer we use `Add`... for mutating APIs, `With...` for non-mutating. No rule, just personal preference.

Reply:
This was at Jasons request for the opposite :p

---
In reply to: [393338348](https://github.com/dotnet/roslyn/pull/42359#discussion_r393338348) [](ancestors = 393338348)